controllers/fluxcd/git-repository-controller.go

```diff
@@ -0,0 +1,171 @@
+	"kubesphere.io/devops/pkg/api/devops/v1alpha3"
+	"kubesphere.io/devops/pkg/utils/k8sutil"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+//+kubebuilder:rbac:groups=devops.kubesphere.io,resources=gitrepositories,verbs=get;list;watch;update;delete
+//+kubebuilder:rbac:groups="",resources=secrets,verbs=get;create;update;patch;delete
+//+kubebuilder:rbac:groups="",resources=events,verbs=create;patch
+//+kubebuilder:rbac:groups="source.toolkit.fluxcd.io",resources=gitrepositories,verbs=get;list;create;update;delete
+
+// GitRepositoryReconciler is the reconciler of the FluxCDGitRepository
+type GitRepositoryReconciler struct {
+	client.Client
+	log      logr.Logger
+	recorder record.EventRecorder
+}
+
+// Reconcile maintains the FluxCDGitRepository against to the GitRepository
+func (r *GitRepositoryReconciler) Reconcile(req ctrl.Request) (result ctrl.Result, err error) {
+	ctx := context.Background()
+	repo := &v1alpha3.GitRepository{}
+
+	if err = r.Get(ctx, req.NamespacedName, repo); err != nil {
+		err = client.IgnoreNotFound(err)
+		return
+	}
+
+	err = r.reconcileFluxGitRepo(repo)
+	return
+}
+
+func (r *GitRepositoryReconciler) reconcileFluxGitRepo(repo *v1alpha3.GitRepository) (err error) {
+	ctx := context.Background()
+	FluxGitRepo := createBareFluxGitRepoObject()
+	// FluxGitRepo's namespace = v1alpha3.GitRepository's namespace
+	// FluxGitRepo's name = "fluxcd-" + v1alpha3.GitRepository's name
+	ns, name := repo.GetNamespace(), getFluxRepoName(repo.GetName())
+
+	if !repo.ObjectMeta.DeletionTimestamp.IsZero() {
+		if err = r.Get(ctx, types.NamespacedName{Namespace: ns, Name: name}, FluxGitRepo); err != nil {
+			err = client.IgnoreNotFound(err)
+		} else {
+			r.log.Info("delete FluxCDGitRepository", "name", FluxGitRepo.GetName())
+			err = r.Delete(ctx, FluxGitRepo)
+		}
+
+		if err == nil {
+			k8sutil.RemoveFinalizer(&repo.ObjectMeta, v1alpha3.GitRepoFinalizerName)
+			err = r.Update(context.TODO(), repo)
+		}
+		return
+	}
+	if k8sutil.AddFinalizer(&repo.ObjectMeta, v1alpha3.GitRepoFinalizerName) {
+		err = r.Update(context.TODO(), repo)
+		if err != nil {
+			return
+		}
+	}
+
+	if err = r.Get(ctx, types.NamespacedName{Namespace: ns, Name: name}, FluxGitRepo); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return
+		}
+		// flux git repo did not existed
+		// create
+		FluxGitRepo := createUnstructuredFluxGitRepo(repo)
+		if err = r.Create(ctx, FluxGitRepo); err != nil {
+			r.recorder.Eventf(FluxGitRepo, v1.EventTypeWarning, "FailedWithFluxCD",
+				"failed to create FluxCDGitRepository, error is: %v", err)
+		}
+	} else {
+		// flux git repo existed
+		// update
+		newFluxGitRepo := createUnstructuredFluxGitRepo(repo)
+		FluxGitRepo.Object["spec"] = newFluxGitRepo.Object["spec"]
+		err = retry.RetryOnConflict(retry.DefaultRetry, func() (err error) {
+			latestGitRepo := createBareFluxGitRepoObject()
+			if err = r.Get(context.Background(), types.NamespacedName{
```

Comment:
How about using `ctx` that in line 63 instead of a new context variable?

Reply:
Thanks a lot for your suggestions. I'll pay attention for situations like these.